org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/managers/ProjectsManager.java

```diff
@@ -127,6 +138,17 @@ public IStatus runInWorkspace(IProgressMonitor monitor) {
 						}
 					}
 					importProjects(addedRootPaths, subMonitor.split(addedRootPaths.size()));
+					Job job = new Job("Register Watchers") {
```

Comment:
we're already running in a job, why add another one here?

Reply:
one issue here is watchers are registered only during server startup/workspace folders update. But if you add resource folders to your pom.xml and reconfigure the project to update classpath settings, new folders won't be watched until next restart or explicit workspace folders update

Reply:
> we're already running in a job, why add another one here?

We have to call JobHelpers.waitForJobsToComplete() in a job to avoid a deadlock

> one issue here is watchers are registered only during server startup/workspace folders update

Fixed.